### PR TITLE
Reuse the same object to write issues across Web Services

### PR DESCRIPTION
**Pull request by jblievremont**
Initial trial to refactor the serialization of objects in Web Services. Medium lived branch: backend modifications must be integrated on client side one by one to avoid breaking UI.

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  private void writeScmAccountsIfNeeded(JsonWriter json, Collection<String> fields, User user) {
+    if (fieldIsWanted(FIELD_SCM_ACCOUNTS, fields)) {
+      json.name(FIELD_SCM_ACCOUNTS)
+        .beginArray()
+        .values(((UserDoc) user).scmAccounts())
+        .endArray();
+    }
+  }
+
+  private boolean fieldIsWanted(String field, @Nullable Collection<String> fields) {
```
Comment (stas-vilchik) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java]: `isFieldWanted` sounds betters, doesn't it?

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  public static final Set<String> FIELDS = ImmutableSet.of(FIELD_NAME, FIELD_EMAIL, FIELD_SCM_ACCOUNTS, FIELD_GROUPS, FIELD_ACTIVE);
+  private static final Set<String> CONCISE_FIELDS = ImmutableSet.of(FIELD_NAME, FIELD_EMAIL, FIELD_ACTIVE);
+
+  private final UserSession userSession;
+
+  public UserWriter(UserSession userSession) {
+    this.userSession = userSession;
+  }
+
+  public void writeFull(JsonWriter json, User user, Collection<String> groups, @Nullable Collection<String> fields) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java]: I would not name this method ```writeFull```. It writes the fields specified as a parameter, for instance:
```
write(json, user, UserWriter.ALL_FIELDS);
write(json, user, UserWriter.CONCISE_FIELDS);
```
Moreover please avoid nullable collections.

Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java]: Next step to support the request param ```f``` is to simply merge the fields, for instance ```write(json, user, Sets.intersection(ALL_FIELDS, requestedFields))```

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.Collection;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.sonar.api.user.User;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.permission.GlobalPermissions;
+import org.sonar.server.user.UserSession;
+import org.sonar.server.user.index.UserDoc;
+
+public class UserWriter {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/UserWriter.java]: ```UserJsonWriter``` ?

**Comment by jblievremont** [comment]
Groups are not present in ``UserDoc`` because they were not needed in the initial implementation of the ``api/users`` Web Services. Now this would probably make sense, to index the group membership information in users - let's discuss IRL.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 4]
```diff
@@ -19,11 +19,14 @@
  */
 package org.sonar.server.issue.ws;
 
+import org.sonar.core.user.GroupDto;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: MINOR Remove this unused import 'org.sonar.core.user.GroupDto'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 1 major
* 1 minor

Details:
* [Make "writeUsers" a "static" method.](https://github.com/SonarSource/sonarqube/blob/f0de4d6b38b18ad7ad1c73c67a8adcfe3bda265b/server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java#L696) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by jblievremont** [comment]
Applied feedback, removed quality flaws and rebased against master. @stas-vilchik, passing to you for update of UI components.

File: server/sonar-server/src/main/java/org/sonar/server/issue/InternalRubyIssueService.java [hunk 5]
```diff
@@ -667,4 +690,94 @@ static SearchOptions toSearchOptions(Map<String, Object> props) {
...
+    if (issue.assignee() != null) {
+      usersByLogin.put(issue.assignee(), userIndex.getByLogin(issue.assignee()));
+    }
+    if (issue.reporter() != null) {
+      usersByLogin.put(issue.reporter(), userIndex.getByLogin(issue.reporter()));
+    }
+    return usersByLogin;
+  }
+
+  private Map<String, ComponentDto> getProjectsByComponentUuid(Collection<ComponentDto> components, Collection<ComponentDto> projects) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/InternalRubyIssueService.java]: MAJOR Make "getProjectsByComponentUuid" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+  public IssueJsonWriter(I18n i18n, Durations durations, UserSession userSession, UserJsonWriter userWriter, IssueActionsWriter actionsWriter) {
+    this.i18n = i18n;
+    this.durations = durations;
+    this.userSession = userSession;
+    this.userWriter = userWriter;
+    this.actionsWriter = actionsWriter;
+  }
+
+  public void write(JsonWriter json, Issue issue, Map<String, User> usersByLogin, Map<String, ComponentDto> componentsByUuid,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java]: MAJOR Method has 8 parameters, which is greater than 7 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+      .prop("severity", issue.severity())
+      .prop("message", issue.message())
+      .prop("line", issue.line())
+      .prop("debt", debt != null ? durations.encode(debt) : null)
+      .prop("reporter", issue.reporter())
+      .prop("author", issue.authorLogin())
+      .prop("actionPlan", actionPlanKey)
+      .prop("creationDate", isoDate(issue.creationDate()))
+      .prop("updateDate", isoDate(updateDate))
+      // TODO Remove as part of Front-end rework on Issue Domain
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    if (file != null) {
+      project = projectsByComponentUuid.get(file.uuid());
+      if (!file.projectUuid().equals(file.moduleUuid())) {
+        subProject = componentsByUuid.get(file.moduleUuid());
+      }
+    }
+    Duration debt = issue.debt();
+    Date updateDate = issue.updateDate();
+
+    json
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/IssueJsonWriter.java]: MAJOR Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 new issues:
* 3 major
* 1 info

File: server/sonar-web/src/main/js/apps/issues/models/issue.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+define([
+  'components/issue/models/issue'
+], function (Issue) {
+
+  return Issue.extend({
+    reset: function (attrs, options) {
+      // TODO remove me soon
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/apps/issues/models/issue.js]: INFO Complete the task associated to this TODO comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3AS1135)

File: server/sonar-web/src/main/js/components/issue/collections/issues.js [hunk 10]
```diff
@@ -26,8 +26,7 @@ define([
       return r.issues.map(function (issue) {
         var component = find(r.components, issue.component),
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/issue/collections/issues.js]: MINOR Add a semicolon at the end of this statement. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=javascript%3ASemicolon)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 new issues:
* 2 major
* 1 minor
* 2 info

**Pull request closed without merging**